### PR TITLE
Add 'Claimed Directories' tab

### PR DESCRIPTION
**Pull request by SkyHaines**

### CODE REVIEW

**Review by mjkw31 — APPROVED** [review]
LGTM.

File: backend/claimstats.go [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+	"encoding/json"
+	"net/http"
+	"time"
+
+	"github.com/wtsi-hgi/backup-plans/db"
+	"github.com/wtsi-hgi/backup-plans/ibackup"
+	"github.com/wtsi-hgi/backup-plans/ruletree"
+)
+
+type RuleStats struct {
```
Comment (sb10) [inline backend/claimstats.go]: Docs for all public things please

**Review by sb10 — CHANGES_REQUESTED** [review]
Test with edge cases. For example, why I look at my page it says the last successful backup was 105666 weeks, and there's nothing in the "Match" column of my single-row table.

See the Jira description for missing features that need implementing: choice of picking team or bom that you belong to. If it's easier, can allow picking any user, team or bom, but default to the logged-in user.

File: backend/claimstats.go [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/wtsi-hgi/backup-plans/db"
+	"github.com/wtsi-hgi/backup-plans/ibackup"
+	"github.com/wtsi-hgi/backup-plans/ruletree"
+)
+
+var ErrNoFilter = errors.New("must provide a user and/or group to filter by")
```
Comment (sb10) [inline backend/claimstats.go]: user group or bom?

File: backend/claimstats.go [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+
+type ruleStats struct {
+	*db.Rule
+	SizeCount
+}
+
+type DirStats struct {
```
Comment (sb10) [inline backend/claimstats.go:47]: docs

File: backend/claimstats.go [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+	return &DirStats{
+		Path:         path,
+		ClaimedBy:    dirSummary.ClaimedBy,
+		Group:        dirSummary.Group,
+		BackupStatus: *sba,
+		RuleStats:    rulestats,
+	}
+}
+
+// generateRuleStats will create a []RuleStats slice for the given directory, containing a RuleStats object for every
```
Comment (sb10) [inline backend/claimstats.go:170]: rewrap

**Review by sb10 — CHANGES_REQUESTED** [review]

File: backend/claimstats.go (hunk 4 above)
Comment (SkyHaines) [reply backend/claimstats.go:170]: Can you elaborate

File: frontend/src/claimstats.ts [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+                users.add(dirStats.ClaimedBy);
+                userList.append(option({ "label": "User: " + dirStats.ClaimedBy }, dirStats.ClaimedBy));
+            };
+
+            page.appendChild(fieldset({ "class": "userclaims", "data-user": dirStats.ClaimedBy, "data-group": dirStats.Group }, [
+                legend({ "class": "claimstats-legend" }, [h2(dirStats.Path), button({
+                    "class": "load-button",
+                    "click": () => load(dirStats.Path).then(() => {
+                        window.scrollTo(0, 0);
+                        document.getElementsByTagName("summary")[0].click();
```
[the remark this answers is not in the document]
Comment (SkyHaines) [reply frontend/src/claimstats.ts:46]: There will always be atleast one summary element and, if there is not, we have bigger issues.

File: backend/claimstats.go [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+	filterUser  bool
+	filterGroup bool
+}
+
+// ClaimStats is an HTTP endpoint that produces a DirStats summary for every claimed directory.
+func (s *Server) ClaimStats(w http.ResponseWriter, r *http.Request) {
+	handle(w, r, s.claimstats)
+}
+
+func (s *Server) claimstats(w http.ResponseWriter, r *http.Request) error { //nolint:funlen
```
Comment (sb10) [inline backend/claimstats.go]: No need for a funlen nolint? This looks trivial to split out with both a for loop and an anonymous goroutine.

File: backend/claimstats.go [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+func (s *Server) filterOutUser(dir *ruletree.DirRules, f filter) bool {
+	return dir.ClaimedBy == "" || (f.filterUser && f.user != dir.ClaimedBy)
+}
+
+// filterGroup will return true if the group/bom does not match the filter.
+func (s *Server) filterOutGroupBom(dir *ruletree.DirRules, f filter) bool {
+	return f.filterGroup && (s.dirGroups[dir.ID()] != f.group && s.dirBoms[dir.ID()] != f.group)
+}
+
+func (s *Server) getFormValues(r *http.Request) filter {
```
Comment (mjkw31) [inline backend/claimstats.go]: Can this method be less generically names, and it should probably be a function as I don't think it uses `s`.

File: backend/claimstats.go [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+
+	return json.NewEncoder(w).Encode(claimstats)
+}
+
+func (s *Server) matchesFilter(dir *ruletree.DirRules, f filter) bool {
+	if !f.filterUser && !f.filterGroup {
+		return false
+	}
+
+	if s.filterOutUser(dir, f) || s.filterOutGroupBom(dir, f) {
```
Comment (mjkw31) [inline backend/claimstats.go]: This can be changed to:

```go
return !s.filterOutUser(dir, f) && !s.filterOutGroupBom(dir, f)
```

**Review by sb10 — CHANGES_REQUESTED** [review]

File: backend/claimstats.go [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+		return false
+	}
+
+	return true
+}
+
+// filterOutUser will return true if the user does not match the filter.
+func (s *Server) filterOutUser(dir *ruletree.DirRules, f filter) bool {
+	return dir.ClaimedBy == "" || (f.filterUser && f.user != dir.ClaimedBy)
```
Comment (mjkw31) [inline backend/claimstats.go]: Given that we're going over a list of claimed directories, is it possible that ClaimedBy is blank?

File: backend/claimstats.go [hunk 10]
```diff
@@ -0,0 +1,225 @@
...
+func (s *Server) filterOutGroupBom(dir *ruletree.DirRules, f filter) bool {
+	return f.filterGroup && (s.dirGroups[dir.ID()] != f.group && s.dirBoms[dir.ID()] != f.group)
+}
+
+func (s *Server) getFormValues(r *http.Request) filter {
+	filterUser := false
+	filterGroup := false
+
+	user := r.FormValue("user")
+	if user != "" {
```
Comment (mjkw31) [inline backend/claimstats.go]:
```go
filterUser := user != ""
```

…same with group below.

File: backend/claimstats.go [hunk 11]
```diff
@@ -0,0 +1,225 @@
...
+		if _, exists := ids[r.ID]; exists || r.ID == 0 {
+			rulestats = append(rulestats, s.generateStatsForRule(r))
+		}
+	}
+
+	return rulestats
+}
+
+func (s *Server) generateStatsForRule(r ruletree.Rule) ruleStats {
+	totalSize := uint64(0)
```
Comment (mjkw31) [inline backend/claimstats.go]: It'd be more idiomatic to write:

```go
var totalSize, totalCount uint64
```

File: backend/server.go [hunk 12]
```diff
@@ -81,6 +86,81 @@ func New(db *db.DB, getUser func(r *http.Request) string, c *config.Config) (*Se
...
+
+	return nil
+}
+
+// UpdateDirMaps will update s.dirGroups and s.dirBoms for the given root.
+func (s *Server) UpdateDirMaps(rootPath string) error {
+	s.rulesMu.Lock()
+	defer s.rulesMu.Unlock()
+
+	mp := s.rootDir.GetMountPoint(rootPath)
```
Comment (mjkw31) [inline backend/server.go]: Is `rootPath` not already the mountpoint?

File: backend/tree.go [hunk 13]
```diff
@@ -50,7 +50,16 @@ var (
 // AddTree adds a tree database, specified by the given file path, to the
 // server, possibly overriding an existing database if they share the same root.
 func (s *Server) AddTree(file string) error {
-	return s.rootDir.AddTree(file)
+	rootPath, err := s.rootDir.AddTree(file)
+	if err != nil {
+		return err
+	}
+
+	if s.dirGroups != nil {
```
Comment (mjkw31) [inline backend/tree.go]: Hmm, this check will need need a read-lock 'round it.

Although, Is it ever non-nil?

File: frontend/src/claimstats.ts [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+    updateClaimStats();
+    base.append(container);
+}
+
+export function updateClaimStats() {
+    clearNode(container);
+    amendNode(container, createClaimStatsSection());
+}
+
+let filter = {
```
Comment (mjkw31) [inline frontend/src/claimstats.ts]: This should be a `const`.

File: frontend/src/claimstats.ts [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+    "user": user,
+    "groupbom": ""
+}
+
+function createClaimStatsSection() {
+    let page = div({ "class": "claimstats-container" });
```
Comment (mjkw31) [inline frontend/src/claimstats.ts]: Also a `const`.

File: frontend/src/claimstats.ts (hunk 14 above)
Comment (mjkw31) [inline frontend/src/claimstats.ts]: This should be a `const`.

File: frontend/src/claimstats.ts (hunk 15 above)
Comment (mjkw31) [inline frontend/src/claimstats.ts]: Also a `const`.

File: backend/server.go [hunk 16]
```diff
@@ -81,6 +86,81 @@ func New(db *db.DB, getUser func(r *http.Request) string, c *config.Config) (*Se
...
+			}
+
+			return err
+		}
+
+		_, gid := dirSummary.IDs()
+		groupname := users.Group(gid)
+		dirGroups[dir.ID()] = groupname
+		dirBoms[dir.ID()] = reverseBomMap[groupname]
+	}
```
[the remark this answers is not in the document]
Comment (SkyHaines) [reply backend/server.go]: LoadDirMaps is only ever called once during initialisation, updates happen via UpdateDirMaps which holds an appropriate lock.

File: ruletree/toplevel.go [hunk 17]
```diff
@@ -150,7 +151,7 @@ func addRules(directoryRules map[string]*DirRules, dirRules []DirRule) ([]string
 	return dirs, nil
 }
 
-func (r *RootDir) getMountPoint(dir string) string {
+func (r *RootDir) GetMountPoint(dir string) string {
```
Comment (mjkw31) [inline ruletree/toplevel.go:156]: This'll need documenting now it's exported.

**Review by mjkw31 — CHANGES_REQUESTED** [review]

File: backend/tree.go (hunk 13 above)
Comment (SkyHaines) [reply backend/tree.go]: It can be technically I believe, because AddTree is called on server start but the dirMaps are not initialised until after all trees are added (i guess it could be two separate functions)

File: backend/server.go (hunk 12 above)
Comment (SkyHaines) [reply backend/server.go]: I don't know tbh, are u saying it is?

Comment (mjkw31) [reply backend/server.go]: It should be (I cannot think what other path would be being returned by `AddTree`, but print to confirm.

File: backend/server.go [hunk 18]
```diff
@@ -81,6 +86,81 @@ func New(db *db.DB, getUser func(r *http.Request) string, c *config.Config) (*Se
...
+
+	s.dirGroups = dirGroups
+	s.dirBoms = dirBoms
+
+	return nil
+}
```
[the remark this answers is not in the document]
Comment (SkyHaines) [reply backend/server.go]: Again, this is only ever called during intialisation after all trees are loaded so no lock needed. UpdateDirMaps does update them and thus has the correct lock.

File: backend/claimstats.go [hunk 19]
```diff
@@ -0,0 +1,229 @@
...
+			if err != nil {
+				return
+			}
+
+			dirSummary.ClaimedBy = s.getClaimed(dir.Path)
+
+			dirStats := s.generateDirStats(dir.Path, dirSummary)
+			channel <- *dirStats
+		}(dir)
+	}
```
[the remark this answers is not in the document]
Comment (SkyHaines) [reply backend/claimstats.go:111]: Considering our usecase limiting the number is not necessary.

File: frontend/src/claimstats.ts [hunk 20]
```diff
@@ -0,0 +1,122 @@
...
+                                    td(formatBytes(BigInt(rule.size))),
+                                    td("" + rule.count)
+                                ])
+                            ]) : tr(td({ "colspan": "4" }, "No rules or unplanned data."))
+                        ])
+                    ])
+                ])
+            ]))
+        }
+        ) : page.appendChild(h2("No claimed directories."))
```
[the remark this answers is not in the document]
Comment (SkyHaines) [reply frontend/src/claimstats.ts]: .map offers more flexibility and keeps the code consistent with the rest of the repo

File: backend/tree.go (hunk 13 above)
Comment (mjkw31) [reply backend/tree.go]: It's probably better to make the map when the struct is created and then only use the `UpdateDirMaps` method; it's mostly duplicating `LoadDirMaps`, and the bits that are used for initialisation can be moved elsewhere.

That way you just always call `UpdateDirMaps` (does that need to be public?) and the check isn't necessary.

File: frontend/src/claimstats.ts [hunk 21]
```diff
@@ -0,0 +1,122 @@
...
+                    "click": () => load(dirStats.Path).then(() => {
+                        window.scrollTo(0, 0);
+                        document.getElementsByTagName("summary")[0].click();
+                    })
+                }, svg([title("Go to"), use({ href: "#goto" })]))]),
+                div([
+                    div({ "class": "claiminfo" }, [
+                        p("Last successful backup: " + (dirStats.BackupStatus.LastSuccess === "0001-01-01T00:00:00Z" ? "Pending" : longAgoStr(dirStats.BackupStatus.LastSuccess))),
+                        p("Failures: " + dirStats.BackupStatus.Failures),
+                        (filter.groupbom !== "" && filter.user === "") ? p("Claimed by: " + dirStats.ClaimedBy) : p()
```
[the remark this answers is not in the document]
Comment (SkyHaines) [reply frontend/src/claimstats.ts]: null results in a syntax error, hence the empty p()

File: frontend/src/claimstats.ts (hunk 20 above)
[the remark this answers is not in the document]
Comment (mjkw31) [reply frontend/src/claimstats.ts]: It could be an `if` statement for the `claimstats.length`, and then a `for const` over the `claimstats`.

Also, would prefer `append` to `appendChild` unless the return value is needed.

Alternatively, it should be something like:

```typescript
        page.replaceChildren(claimstats.length > 0 ? claimstats.map((dirStats) => {
            if (!users.has(dirStats.ClaimedBy)) {
                users.add(dirStats.ClaimedBy);
                userList.append(option({ "label": "User: " + dirStats.ClaimedBy }, dirStats.ClaimedBy));
            };

            return fieldset({ "class": "userclaims", "data-user": dirStats.ClaimedBy, "data-group": dirStats.Group }, [
                // blah
            ]);
} : tr(td({ "colspan": "4" }, "No rules or unplanned data.")));
```

You won't need to explicitly remove the spinner that way.

File: frontend/src/claimstats.ts (hunk 21 above)
[the remark this answers is not in the document]
Comment (mjkw31) [reply frontend/src/claimstats.ts]: It should not be `null`, but an empty array `[]`.

**Review by sb10 — APPROVED** [review]
